File: src/types/reader.ts
export type PlaybackState = "idle" | "loading" | "playing" | "paused";

// VoiceId is a string identifier for Kokoro voices (e.g., "af_heart", "am_adam")
export type VoiceId = string;

export type AudioTrack = {
  id: string;
  title: string;
  href: string;
  url?: string; // Optional - loaded lazily
  duration?: number;
  _loading?: boolean; // Internal flag to track loading state
};

export type AudioSyncSegment = {
  textElementId: string;
  chapterHref: string;
  audioTrackHref: string;
  clipBegin: number; // seconds
  clipEnd: number; // seconds
};

export type AudioSyncMap = {
  segments: AudioSyncSegment[];
  // Map from audioTrackHref + time to segment index for quick lookup (optional, not used by findCurrentAudioSegment)
  lookup?: Map<string, number>;
};

export type BookAudioState = {
  currentTrackId: string;
  currentTrackHref: string;
  currentTrackIndex: number;
  currentTimeSeconds: number;
  updatedAt: string;
};

export type Chapter = {
  id: string;
  title: string;
  contentHtml?: string; // Optional - loaded lazily
  plainText?: string; // Optional - loaded lazily
  order: number;
  href: string;
  wordCount?: number;
  estimatedPageCount?: number;
  _loading?: boolean; // Internal flag to track loading state
};

export type BookProgress = {
  currentChapterId: string;
  currentChapterHref: string;
  currentChapterIndex: number;
  currentChapterElementId?: string | null;
  currentChapterElementIndex?: number | null;
  currentChapterScrollTop: number;
  currentChapterScrollHeight: number;
  currentChapterClientHeight: number;
  chapterProgressPercent: number;
  /** Overall book progress across all chapters (0.0 to 1.0) */
  bookProgressPercent: number;
  updatedAt: string;
};

export type Book = {
  id: string;
  title: string;
  author: string;
  chapters: Chapter[];
  coverUrl?: string;
  sourcePath: string;
  contentHash?: string;
  publisher?: string;
  publishedYear?: string;
  subjects?: string[];
  fileSizeBytes?: number;
  audioTracks: AudioTrack[];
  audioState?: BookAudioState;
  audioSyncMap?: AudioSyncMap;
  progress?: BookProgress;
  pageCount?: number;
  conversionStarted?: boolean;
  completedChapters?: string[];
};

export type NavItem = {
  href: string;
  label?: string;
  subitems?: NavItem[];
};

export type ReaderTheme = "light" | "dark" | "system";

export type ReaderFont =
  | "merriweather"
  | "inter"
  | "lora"
  | "firaMono"
  | "atkinson";

export type ReaderContentPadding = "compact" | "comfortable" | "spacious";

export type ReaderFontSize = "small" | "medium" | "large" | "xlarge";

export type ReaderPreferences = {
  theme: ReaderTheme;
  fontFamily: ReaderFont;
  contentPadding: ReaderContentPadding;
  fontSize: ReaderFontSize;
};

File: src-frontend/src/types/reader.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum PlaybackState {
    Idle,
    Loading,
    Playing,
    Paused,
}

/// VoiceId is a string identifier for Kokoro voices (e.g., "af_heart", "am_adam")
pub type VoiceId = String;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AudioTrack {
    pub id: String,
    pub title: String,
    pub href: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub url: Option<String>, // Optional - loaded lazily
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration: Option<f64>,
    pub order: usize, // Order matches corresponding chapter order
    #[serde(skip)]
    pub loading: bool, // Internal flag to track loading state
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AudioSyncSegment {
    pub text_element_id: String,
    pub chapter_href: String,
    pub audio_track_href: String,
    pub clip_begin: f64, // seconds
    pub clip_end: f64, // seconds
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AudioSyncMap {
    pub segments: Vec<AudioSyncSegment>,
    // Map from audioTrackHref + time to segment index for quick lookup (optional, not used by findCurrentAudioSegment)
    #[serde(skip)]
    pub lookup: Option<std::collections::HashMap<String, usize>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct BookAudioState {
    pub current_track_id: String,
    pub current_track_href: String,
    pub current_track_index: usize,
    pub current_time_seconds: f64,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Chapter {
    pub id: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_html: Option<String>, // Optional - loaded lazily
    #[serde(skip_serializing_if = "Option::is_none")]
    pub plain_text: Option<String>, // Optional - loaded lazily
    pub order: usize,
    pub href: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub word_count: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub estimated_page_count: Option<usize>,
    #[serde(skip)]
    pub loading: bool, // Internal flag to track loading state
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct BookProgress {
    pub current_chapter_id: String,
    pub current_chapter_href: String,
    pub current_chapter_index: usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_chapter_element_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_chapter_element_index: Option<usize>,
    pub current_chapter_scroll_top: f64,
    pub current_chapter_scroll_height: f64,
    pub current_chapter_client_height: f64,
    pub chapter_progress_percent: f64,
    /// Overall book progress across all chapters (0.0 to 1.0)
    pub book_progress_percent: f64,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Book {
    pub id: String,
    pub title: String,
    pub author: String,
    pub chapters: Vec<Chapter>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cover_url: Option<String>,
    pub source_path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_hash: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub publisher: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub published_year: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subjects: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_size_bytes: Option<u64>,
    pub audio_tracks: Vec<AudioTrack>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audio_state: Option<BookAudioState>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audio_sync_map: Option<AudioSyncMap>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub progress: Option<BookProgress>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub page_count: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conversion_status: Option<String>, // "notStarted" | "started" | "done"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_chapters: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub voice_id: Option<VoiceId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_words: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub words_processed: Option<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_opened_time: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct NavItem {
    pub href: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subitems: Option<Vec<NavItem>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ReaderTheme {
    Light,
    Dark,
    System,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ReaderFont {
    Merriweather,
    Inter,
    Lora,
    FiraMono,
    Atkinson,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ReaderContentPadding {
    Compact,
    Comfortable,
    Spacious,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ReaderFontSize {
    Small,
    Medium,
    Large,
    XLarge,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ReaderPreferences {
    pub theme: ReaderTheme,
    pub font_family: ReaderFont,
    pub content_padding: ReaderContentPadding,
    pub font_size: ReaderFontSize,
}
